### PR TITLE
Handle obsolete bundle versions in IntegrationTest (#2003)

### PR DESCRIPTION
**Pull request by amarjandu**

### CODE REVIEW

File: test/integration_test.py [hunk 1]
```diff
@@ -326,14 +335,27 @@ def _prune_test_bundles(self, bundle_fqids: List[BundleFQID], max_bundles: int)
...
     def _assert_bundles_are_indexed(self, entity_type: str):
         timeout = 600
         indexed_fqids = set()
-        expected_fqids = self.expected_fqids
+        expected_fqids = self._filter_superseded_bundle_versions(set(self.expected_fqids))
+        num_expected_bundles = len(expected_fqids)
+        superseded_bundles = self.expected_fqids - expected_fqids
+        if superseded_bundles:
+            log.debug('Inspections for bundles: %r were removed due to indexing a superseded version of the bundle',
+                      superseded_bundles)
```
Comment (danielsotirhos) [inline test/integration_test.py:358]: How do you feel about moving lines 355-358 into `_filter_superseded_bundle_versions()`? It'd help keep `_assert_bundles_are_indexed()` focused on it's purpose.

Comment (amarjandu) [reply test/integration_test.py:358]: i'm up for that modification.

File: test/integration_test.py [hunk 2]
```diff
@@ -326,14 +335,28 @@ def _prune_test_bundles(self, bundle_fqids: List[BundleFQID], max_bundles: int)
                 break
         return filtered_bundle_fqids
 
+    def _filter_superseded_bundle_versions(self, bundle_fqids: Set[BundleFQID]) -> Set[BundleFQID]:
+        filtered_fqids = {
```
Comment (hannes-ucsc) [inline test/integration_test.py:339]: No need for grouping and sorting if you're producing a dict. The dict is a grouping data structure in itself. Just use a `defaultdict(list)` and successively append to the values. Then use max to get the latest version. And try to avoid having to reconstruct the BundleFQID tuples by making the values lists of full BundleFQIDs and not just versions.

File: test/integration_test.py [hunk 3]
```diff
@@ -306,6 +309,12 @@ def _validate_fastq_content(self, file_uuid: str, content: bytes):
     def _prepare_notifications(self) -> Dict[BundleFQID, JSON]:
         bundle_fqids = self.azul_client.list_bundles()
         bundle_fqids = self._prune_test_bundles(bundle_fqids, self.max_bundles)
+        bundle_fqids.pop()
+        bundle_fqids.pop()
+        bundle_fqids.append(
+            BundleFQID(uuid='d6eadc47-4b74-418b-a3f8-5642e8814a4a', version='2019-07-26T150634.198817Z'))
```
Comment (hannes-ucsc) [inline test/integration_test.py:315]: This can't work as intended. The TDR snapshot doesn't have these bundles. You may need to include code in the drop commit that switches back to the DSS plugin. Then simply ping the prefix and the pruning seed to the same ones that were used in the build I reference in the ticket.

File: test/integration_test.py (hunk 2 above)
Comment (hannes-ucsc) [inline test/integration_test.py:339]:
```suggestion
        bundle_fqids_by_uuid = {
```

File: test/integration_test.py [hunk 4]
```diff
@@ -326,14 +335,28 @@ def _prune_test_bundles(self, bundle_fqids: List[BundleFQID], max_bundles: int)
                 break
         return filtered_bundle_fqids
 
+    def _filter_superseded_bundle_versions(self, bundle_fqids: Set[BundleFQID]) -> Set[BundleFQID]:
+        filtered_fqids = {
+            uuid_: [fqid.version for fqid in items]
+            for uuid_, items in groupby(sorted(bundle_fqids, key=attrgetter("uuid", "version")),
+                                        key=attrgetter("uuid"))
+        }
+        filtered_fqids_set = set(BundleFQID(k, v[-1]) for k, v in filtered_fqids.items())
```
Comment (hannes-ucsc) [inline test/integration_test.py:344]: latest_bundle_fqids

File: test/integration_test.py [hunk 5]
```diff
@@ -326,14 +335,28 @@ def _prune_test_bundles(self, bundle_fqids: List[BundleFQID], max_bundles: int)
...
+        superseded_bundles = bundle_fqids - filtered_fqids_set
```
Comment (hannes-ucsc) [inline test/integration_test.py:345]:
```suggestion
        superseded_bundle_fqids = bundle_fqids - filtered_fqids_set
```

File: test/integration_test.py [hunk 6]
```diff
@@ -326,14 +335,28 @@ def _prune_test_bundles(self, bundle_fqids: List[BundleFQID], max_bundles: int)
...
+        if superseded_bundles:
+            log.debug('Inspections for bundles: %r were removed due to indexing a superseded version of the bundle',
```
Comment (hannes-ucsc) [inline test/integration_test.py:347]:
```suggestion
            log.debug('Ignoring superseded bundle versions %r',
```

File: test/integration_test.py (hunk 3 above)
Comment (hannes-ucsc) [reply test/integration_test.py:315]: s/ping/pin

**Comment by hannes-ucsc** [comment]
Note that I squashed and force-pushed your branch.

File: test/integration_test.py (hunk 3 above)
Comment (hannes-ucsc) [reply test/integration_test.py]: Just noticed a bug in the pruning method that could cause a non-deterministic random order even if the same pruning seed is used. It all depends on whether the input list of FQIDs to _prune_test_bundles is ordered deterministically.

Comment (hannes-ucsc) [reply test/integration_test.py]: If it is, you should be ok. Otherwise you may not be able to reproduce the issue by pinning the seed.

Comment (amarjandu) [reply test/integration_test.py]: I was running the IT locally when you posted the message, it looks like it is able to reproduce the same set of bundles.
```
2020-07-24 12:47:31,589 DEBUG   MainThread: Ignoring superseded bundle versions {BundleFQID(uuid='d6eadc47-4b74-418b-a3f8-5642e8814a4a', version='2019-07-26T150634.198817Z')}
2020-07-24 12:47:31,589 DEBUG   MainThread: Expecting bundles [BundleFQID(uuid='d6015595-a613-4211-804e-a3ac78849ed6', version='2019-11-04T202827.896050Z'), BundleFQID(uuid='d607a428-67a8-4c0e-960c-29044848628b', version='2019-05-30T212106.375000Z'), BundleFQID(uuid='d60fee46-3309-428e-b4f9-9b893ec705dd', version='2019-08-01T200147.471022Z'), BundleFQID(uuid='d611f11b-5512-400a-8d42-edfbe5886620', version='2019-08-01T200147.532482Z'), BundleFQID(uuid='d618be39-3f63-4ee1-86cf-edb4d9fb7bc3', version='2019-08-01T200147.312113Z'), BundleFQID(uuid='d61a489f-5f4e-4070-8b55-3090b3059f41', version='2019-08-01T200147.513600Z'), BundleFQID(uuid='d61f67c2-d641-43c1-84fe-1e83ab169715', version='2019-08-01T200147.316820Z'), BundleFQID(uuid='d61fb632-f048-4c88-b237-525234edc3b1', version='2019-08-01T200147.546712Z'), BundleFQID(uuid='d6205183-1e8a-4154-8735-472759ea5fe3', version='2019-08-01T200147.932077Z'), BundleFQID(uuid='d622c263-8fa5-4a94-9ef3-d38b96ff0815', version='2019-08-01T200147.081977Z'), BundleFQID(uuid='d62cb03f-20eb-4e79-922e-74d5be70658f', version='2019-08-01T200147.542444Z'), BundleFQID(uuid='d6358e7b-08f2-42b3-bbe6-86274f3651d0', version='2019-08-01T200146.977855Z'), BundleFQID(uuid='d6379d54-6eb2-4fd8-a268-5bcf94d232d9', version='2019-08-01T200147.129743Z'), BundleFQID(uuid='d638ab7d-54ed-4b95-8e72-1642f6841a8c', version='2019-08-01T200147.506013Z'), BundleFQID(uuid='d63c0643-c4be-4339-bb3a-383720dac3ea', version='2019-05-18T004924.380000Z'), BundleFQID(uuid='d63d76e3-ab61-498a-8337-aaf6ae7739ae', version='2019-08-01T200147.905280Z'), BundleFQID(uuid='d63de872-d480-4eed-8de9-82f839a2bd09', version='2019-08-01T200147.886722Z'), BundleFQID(uuid='d6428a0b-3dee-43e5-8855-ddc7111ebbe5', version='2019-05-14T083819.443000Z'), BundleFQID(uuid='d643b817-0f17-41da-923c-037cacfc8def', version='2019-08-01T200147.853548Z'), BundleFQID(uuid='d6445ff5-f518-43cd-8828-d8c9a64c37f9', version='2019-05-17T160538.682000Z'), BundleFQID(uuid='d64868bf-d4e1-465c-8386-f60116e2b568', version='2019-08-01T200147.066251Z'), BundleFQID(uuid='d651c9d0-5cf4-4e16-b5e7-2165af44cbaa', version='2019-08-01T200147.350765Z'), BundleFQID(uuid='d651ebf1-c8a6-4662-914f-90fe134fc1f9', version='2019-08-01T200147.138795Z'), BundleFQID(uuid='d6592eb5-76a7-4750-a9ce-4cf6ed271ac5', version='2019-11-04T202252.273932Z'), BundleFQID(uuid='d65c8a52-6d55-44e2-b6fd-8e4b30b366c2', version='2019-08-01T200147.216796Z'), BundleFQID(uuid='d6607afd-c09b-4130-986f-dd9e0fab4dfb', version='2019-08-01T200147.123307Z'), BundleFQID(uuid='d661a628-c76b-470c-99ef-f57a617a50e7', version='2019-08-01T200147.090777Z'), BundleFQID(uuid='d6620d99-9824-487f-b5af-badbd7c05138', version='2019-08-01T200147.237051Z'), BundleFQID(uuid='d663cfc7-60d3-4c8c-85a0-ce6340830761', version='2019-08-01T200146.958977Z'), BundleFQID(uuid='d6654db5-3aa2-48ed-b5f8-0cab2debf80e', version='2019-08-01T200147.447392Z'), BundleFQID(uuid='d66b5728-c9af-453e-972b-fdf452cb1fad', version='2019-08-01T200147.774404Z'), BundleFQID(uuid='d676394a-0775-426e-be17-fe8a1558fd88', version='2019-08-01T200146.971255Z'), BundleFQID(uuid='d678800c-5932-4fa2-97f1-7c7a8cffa51a', version='2019-08-01T200147.209848Z'), BundleFQID(uuid='d67947f9-dca0-4e97-a440-a7cb19b08320', version='2019-08-01T200147.136506Z'), BundleFQID(uuid='d67ba02e-e3bf-4725-bcab-e478ef0a86e9', version='2019-08-01T200147.825099Z'), BundleFQID(uuid='d67ec13d-d0e4-4a9c-afef-22e0a748542d', version='2019-08-01T200146.946603Z'), BundleFQID(uuid='d6862c08-4940-4096-84c7-de08cc8bb05a', version='2019-05-18T195707.873000Z'), BundleFQID(uuid='d688f5ab-0f1d-4658-a4b6-baf46ca8264b', version='2019-08-01T200147.944184Z'), BundleFQID(uuid='d68c5283-354f-44a9-865d-86a3a0f04e1d', version='2019-08-01T200147.618435Z'), BundleFQID(uuid='d68c5b97-aae9-4286-8e25-e5dc878a514f', version='2019-05-15T222432.579000Z'), BundleFQID(uuid='d68e57f8-b6ba-4f20-90ea-e8c0b724a442', version='2019-08-01T200147.871049Z'), BundleFQID(uuid='d694849b-a434-4d1c-8684-6a2c3ccb1353', version='2019-08-01T200147.521563Z'), BundleFQID(uuid='d69b1c2c-e50b-4b0a-8feb-aab74a561737', version='2019-08-01T200147.363780Z'), BundleFQID(uuid='d6b49af8-8df7-432c-b953-cd58d35d9224', version='2019-08-01T200147.279571Z'), BundleFQID(uuid='d6c12490-f566-4197-94fe-cc122b652614', version='2019-08-01T200147.830041Z'), BundleFQID(uuid='d6c33cce-41f9-43b8-9daa-494831e9e8f0', version='2019-05-17T160538.703000Z'), BundleFQID(uuid='d6c3efcf-5379-44d8-8bbb-bb82c1d38db1', version='2019-08-01T200147.482687Z'), BundleFQID(uuid='d6c54308-4f94-46a6-aad9-42f120803c5a', version='2019-08-01T200147.669395Z'), BundleFQID(uuid='d6cb2ab6-f34e-4d4b-b0a1-9b7d292ce8c8', version='2019-08-01T200147.256022Z'), BundleFQID(uuid='d6d45a61-1d16-4875-9876-8a11dd737e8d', version='2019-08-01T200147.598769Z'), BundleFQID(uuid='d6d58ade-a35e-46a5-8f4e-329cb049b58a', version='2019-08-01T200147.128201Z'), BundleFQID(uuid='d6d5d68a-8206-469f-8849-ad6597ff274a', version='2019-08-01T200147.427528Z'), BundleFQID(uuid='d6dd0346-cabc-4c0d-9733-4d82235fd447', version='2019-08-01T200147.300680Z'), BundleFQID(uuid='d6e04d24-b51b-42a2-87cb-cb6916e3d354', version='2019-05-18T012714.719000Z'), BundleFQID(uuid='d6e1ebc2-cf91-4389-a77f-29464f5ecae8', version='2019-05-16T154904.483000Z'), BundleFQID(uuid='d6e20c2d-f316-459e-b969-a4191657d92d', version='2019-05-14T105020.492000Z'), BundleFQID(uuid='d6e5ca2c-1c8e-42e8-9221-92b7e2b813fd', version='2019-08-01T200147.938840Z'), BundleFQID(uuid='d6eacdee-08ab-40f8-be3d-8933b819b7ba', version='2019-08-01T200146.964873Z'), BundleFQID(uuid='d6eadc47-4b74-418b-a3f8-5642e8814a4a', version='2019-10-04T094331.956635Z'), BundleFQID(uuid='d6ed2145-8263-4832-b570-12db36ee2644', version='2019-08-01T200147.096542Z'), BundleFQID(uuid='d6f34c0c-86f4-4a6f-98e2-dd9d73fe1c67', version='2019-08-01T200147.078781Z'), BundleFQID(uuid='d6f5bb2a-793e-4933-b38b-f95bb0902c0a', version='2019-08-01T200147.073231Z'), BundleFQID(uuid='d6fd6ff5-43c7-441c-984f-9cf6bcecdf95', version='2019-08-01T200146.968900Z')] 
2020-07-24 12:47:34,771 INFO    MainThread: Detected 63 of 63 bundles in 235 hits for entity type files on try #0.
```

Comment (amarjandu) [reply test/integration_test.py]: I was able to get an IT to fail consistently earlier this morning. https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/snippets/3
This was prior to rebasing on to the current `develop` branch, it looks like now my `indexer service account` needs to be authorize to access the TDR table. I need to confirm that the failures still occur, as it appears that our pruning function has changed slightly.

Comment (amarjandu) [reply test/integration_test.py]: https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/snippets/4 unable to generate bundle list with the problematic bundles we saw. This problem came after the refactoring with changes that were made to the _prune_test_bundles in the IT.

**Comment by amarjandu** [comment]
Got the superseded bundle to appear after further constraining the prefix to be indexed. 
https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/snippets/5

drop! commit will not handle the superseded bundles and fail as desired.
the fixed commit will produced a debug statement like:
`Ignoring superseded bundle versions {BundleFQID(uuid='d6eadc47-4b74-418b-a3f8-5642e8814a4a', version='2019-07-26T150634.198817Z')}`

File: test/integration_test.py (hunk 2 above)
Comment (amarjandu) [reply test/integration_test.py]: i ended up not needing this suggestion.

**Comment by amarjandu** [comment]
> Note that I squashed and force-pushed your branch.

I accidentally pushed over those changes last week, i cleaned up the commits to just have the drop commit and the commit that fixes the issue.

File: deployments/sandbox/environment.py [hunk 7]
```diff
@@ -26,14 +26,15 @@ def env() -> Mapping[str, Optional[str]]:
         # environment.local.py right next to this file and make your changes there.
         # Settings applicable to all environments but specific to you go into
         # environment.local.py at the project root.
-        
+
```
Comment (hannes-ucsc) [inline deployments/sandbox/environment.py:29]: Remove this change from drop! commit.

File: test/integration_test.py [hunk 8]
```diff
@@ -101,10 +101,7 @@ class IndexingIntegrationTest(IntegrationTestCase, AlwaysTearDownTestCase):
     def setUpClass(cls) -> None:
         super().setUpClass()
         # FIXME: https://github.com/DataBiosphere/azul/issues/2023
-        cls.bundle_uuid_prefix = 'a0' or ''.join([
```
Comment (hannes-ucsc) [inline test/integration_test.py:104]: Minimize the change like we talked about.

File: test/integration_test.py [hunk 9]
```diff
@@ -347,13 +344,13 @@ def _prune_test_bundles(self,
                             bundle_fqids: Sequence[BundleFQID],
                             max_bundles: int
                             ) -> List[BundleFQID]:
-        seed = self.pruning_seed
+        seed = 5245369912902346391
```
Comment (hannes-ucsc) [inline test/integration_test.py:350]: Pin where self.pruning_seed is assigned.

File: test/integration_test.py [hunk 10]
```diff
@@ -347,13 +344,13 @@ def _prune_test_bundles(self,
...
         log.info('Selecting %i bundles with projects, out of %i candidates, using random seed %i.',
                  max_bundles, len(bundle_fqids), seed)
         random_ = random.Random(x=seed)
         # The same seed should give same random order so we need to have a
         # deterministic order in the input list.
-        bundle_fqids = sorted(bundle_fqids)
+        bundle_fqids = sorted(bundle_fqids, reverse=True)
```
Comment (hannes-ucsc) [inline test/integration_test.py:356]: Remove if possible or confirm in a comment that it is necessary.

File: test/integration_test.py [hunk 11]
```diff
@@ -363,11 +366,22 @@ def _prune_test_bundles(self,
                 break
         return filtered_bundle_fqids
 
+    @staticmethod
+    def _filter_superseded_bundle_versions(bundle_fqids: Set[BundleFQID]) -> Set[BundleFQID]:
+        latest_bundle_fqids = set(max(bundle_fqid, key=attrgetter("version"))
```
Comment (hannes-ucsc) [inline test/integration_test.py:371]: This doesn't quite address my concerns:

https://github.com/DataBiosphere/azul/pull/2042#discussion_r459634915

I would like you to attempt the dict based solution. We can then compare which one is easier to undestand.

**Comment by amarjandu** [comment]
reindex was kicked off early this morning, unable to run integration tests without errors. 
waiting for the reindex to complete before testing dictionary approach.

**Comment by amarjandu** [comment]
Locally run integration-test handed the superseded bundled as expected with the dictionary based approach.

**Comment by hannes-ucsc** [comment]
I'll boost the estimate a bit.

**Comment by amarjandu** [comment]
> It just occurred to me that there alread is very similar functionality in _filter_obsolete_bundle_versions

It seems that recalling that the functionality exists within the code base is dependent on:
1. knowing that the functionality exists already.
2. a by-product of how the code is designed. 

I was not aware of the functionality, and have not used the AzulClient code before. I should have searched the entire codebase to see if there was something already made. I guess the best way to improve the ability to recall is simply more exposure to parts of the system.

**Comment by hannes-ucsc** [comment]
No worries. I wrote that function and I forgot it existed.

File: src/azul/azulclient.py [hunk 12]
```diff
@@ -358,6 +326,38 @@ def wait_for_indexer(self, **kwargs):
         self.queues.wait_for_queue_level(empty=True, **kwargs)
 
 
+def filter_obsolete_bundle_versions(bundle_fqids: Iterable[BundleFQID]) -> List[BundleFQID]:
```
Comment (hannes-ucsc) [inline src/azul/azulclient.py:329]: Why did you move this method? It makes the diff unnecessarily large so it's hard to detect if you actually modified the method.

Comment (amarjandu) [reply src/azul/azulclient.py]: I was confused by `We need to make it public and remove the leading _ from the name.` I had removed the function from the class entirely.

I amended the changes to the method and have confirmed that the IT works as expected.
Just waiting on Travis now.

File: src/azul/azulclient.py [hunk 13]
```diff
@@ -249,24 +249,24 @@ def _do_remote_index(self, message: JSON) -> None:
             num_messages += len(batch)
         logger.info('Successfully queued %i notification(s) for prefix %s', num_messages, self.prefix)
 
-    @classmethod
-    def _filter_obsolete_bundle_versions(cls, bundle_fqids: Iterable[BundleFQID]) -> List[BundleFQID]:
+    @staticmethod
```
Comment (hannes-ucsc) [inline src/azul/azulclient.py:252]: See CONTRIBUTIONS guide on `@staticmethod`

File: deployments/sandbox/environment.py [hunk 14]
```diff
@@ -26,14 +26,14 @@ def env() -> Mapping[str, Optional[str]]:
         # environment.local.py right next to this file and make your changes there.
         # Settings applicable to all environments but specific to you go into
         # environment.local.py at the project root.
-
```
Comment (hannes-ucsc) [inline deployments/sandbox/environment.py:29]: This empty line should remain.

File: deployments/sandbox/environment.py [hunk 15]
```diff
@@ -26,14 +26,14 @@ def env() -> Mapping[str, Optional[str]]:
...
         # environment.local.py at the project root.
-        
         'AZUL_DEPLOYMENT_STAGE': 'sandbox',
         
         'AZUL_DSS_ENDPOINT': 'https://dss.data.humancellatlas.org/v1',
         'AZUL_DSS_DIRECT_ACCESS': '1',
         'AZUL_DSS_DIRECT_ACCESS_ROLE': 'arn:aws:iam::109067257620:role/azul-sc',
         'AZUL_SUBSCRIBE_TO_DSS': '0',
 
+        'AZUL_REPOSITORY_PLUGIN': 'dss',
```
Comment (hannes-ucsc) [inline deployments/sandbox/environment.py:36]: And this variable isn't used any more.

**Comment by hannes-ucsc** [comment]
Finally got it to fail on the drop! commit as expected: https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/5003

Now pushing the main commits.

**Comment by hannes-ucsc** [comment]
Passes with all commits: https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/5013

Removing drop! commit.

**Comment by hannes-ucsc** [comment]
For demo, just show the Gitlab builds failing and succeeding and explain the thinking behind the validation technique.
